build_catalog.sh

```diff
@@ -1,105 +1,186 @@
 
-[ $count -gt 0 ] && exit 1
-
-num_commits=$(git rev-list $(git rev-list --max-parents=0 HEAD)..HEAD --count)
-current_commit=$(git rev-parse --short=7 HEAD)
-version="0.1.$num_commits-$current_commit"
-opm_version="1.14.0"
-
-# Login to docker
-docker_conf="$PWD/.docker"
-mkdir -p "$docker_conf"
-docker_cmd="docker --config=$docker_conf"
-
-$docker_cmd login -u="$QUAY_USER" -p="$QUAY_TOKEN" quay.io
-
-# Find the CSV version from the previous bundle
-log "Pulling latest bundle image $BUNDLE_IMAGE"
-$docker_cmd pull $BUNDLE_IMAGE:latest && exists=1 || exists=0
-
-if [ $exists -eq 1 ]; then
-  log "Extracting previous version from bundle image"
-  docker create --name="tmp_$$" $BUNDLE_IMAGE:latest sh
-  tmp_dir=$(mktemp -d -t sa-XXXXXXXXXX)
-  pushd $tmp_dir
-    docker export tmp_$$ | tar -xf -
-    prev_version=`find . -name *.clusterserviceversion.* | xargs cat - | python3 -c 'import sys,yaml; print(yaml.safe_load(sys.stdin.read())["metadata"]["name"])'`
-    if [[ "$prev_version" == "" ]]; then
-      log "Unable to find previous bundle version"
-      exit 1
+[[ $count -gt 0 ]] && exit 1
+
+# We shouln't need set -u as we have checked for the vars before
+# but let's play safe
+set -euo pipefail
+
+# General vars to modify behaviour of the script
+DRY_RUN=${DRY_RUN:-true}
+DELETE_TEMP_DIR=${DELETE_TEMP_DIR:-true}
+REMOVE_UNDEPLOYED=${REMOVE_UNDEPLOYED:-false}
+
+temp_dir=$(mktemp -d)
+[[ "$DELETE_TEMP_DIR" == "true" ]] && trap 'rm -rf $temp_dir' EXIT
+
+engine_cmd="$CONTAINER_ENGINE --config=$CONFIG_DIR"
+
+# clone bundle repo containing current version
+saas_operator_dir_base="$temp_dir/saas-operator-dir"
+bundle_versions_file="$saas_operator_dir_base/$OPERATOR_NAME/${OPERATOR_NAME}-versions.txt"
+bundle_versions_package_file="$saas_operator_dir_base/$OPERATOR_NAME/${OPERATOR_NAME}.package.yaml"
+
+log "Cloning $BUNDLE_VERSIONS_REPO"
+if [[ -n "${APP_SRE_BOT_PUSH_USER:-}" && "${APP_SRE_BOT_PUSH_TOKEN:-}" ]]; then
+    bundle_versions_repo_url="https://${APP_SRE_BOT_PUSH_USER}:${APP_SRE_BOT_PUSH_TOKEN}@$BUNDLE_VERSIONS_REPO"
+else
+    bundle_versions_repo_url="https://$BUNDLE_VERSIONS_REPO"
+fi
+
+git clone --branch "$BRANCH_CHANNEL" "$bundle_versions_repo_url" "$saas_operator_dir_base"
+
+prev_operator_version=""
+removed_versions=""
+if [[ -f "$bundle_versions_file" ]]; then
+    log "$bundle_versions_file exists. We'll use to determine current version"
+    if [[ "$REMOVE_UNDEPLOYED" == "true" ]]; then
+        log "Checking if we have to remove any versions more recent than deployed hash"
+        # TODO: Move this out of here
+        deployed_hash=$(
+            curl -s "https://gitlab.cee.redhat.com/service/app-interface/-/raw/master/data/services/deployment-validation-operator/cicd/saas.yaml" | \
+                docker run --rm -i quay.io/app-sre/yq -r '.resourceTemplates[]|select(.name="deployment-validation-operator").targets[]|select(.namespace["$ref"]=="/openshift/app-sre-stage-01/namespaces/app-sre-dvo-per-cluster.yml")|.ref'
```

Comment:
This part is going to be difficult, indeed. I didn't put lots of thought here apart from that `TODO` 😂 

> Perhaps if we made the new deployment-validation-operator-versions.txt a deployment-validation-operator-versions.yaml instead, we could embed the necessary mappings therein...?

I don't understand this well. Do you propose that we add some sort of configuration to the saas repo to help us generalize this part? I'm not sure that I would like to place it in the same file we're overwriting each time but since that belongs basically to the deployment part, it could be an interesting idea to have some sort of config that helps us with this section...

Reply:
Just spitballing. If a given element in the list of versions truly doesn't need to provide any more information than it already does, then it probably makes more sense to put that config in a separate file, as (I think) you're suggesting.

Reply:
I'm actually wondering if we need this logic at all.  If we only ever update the versions file with successful build information then what is actually deployed on the cluster isn't relevant.  All we need to make sure to maintain is the upgrade chain for the catalog.  We shouldn't need to worry about deleting versions newer than what is currently deployed in that case.

Reply:
We currently keep all versions for staging and a clean build tree for production. That's the reason behind the `REMOVED_DEPLOYED` logic

Reply:
I guess I don't see the point in keeping this metadata for stage and production.  The key goal, afaik, for this data is so that we always know what the upgrade chain is and we can re-create it if needed.  To that end, any bundle that is built but is never placed into a catalog is irrelevant.  It will never be deployed to a cluster, but the docker image will still be in quay if anyone really wants to poke at it.  I thought we were trying to solve a build issue, not a deployment issue.  The deployment side seems better tracked in app-interface since it will actually be doing the deployments.

What are we gaining by keeping this logic?

Reply:
> Right now this is hardcoded to be specific to `app-sre-stage-01`. Presumably at some point there will need to be a correlation between the stanza we choose from the saas.yaml and the `$BRANCH_CHANNEL` variable.
> 
> For purposes of generalizing this, it's also going to suck that there's not a more deterministic way to map those together. And when there's more than one of each environment (e.g. there are currently two integration, three staging, and four production environments for OSD) we'll essentially be choosing one at random :(

Not sure if it's serendipity or if @maorfr heard about our problem, but have a look at https://github.com/openshift/aws-account-operator/pull/493

Reply:
> Not sure if it's serendipity or if @maorfr heard about our problem, but have a look at openshift/aws-account-operator#493

I just read it this morning and planned to go that route. Thanks for the pointer anyway

Reply:
i just caught this thread: https://coreos.slack.com/archives/CCRND57FW/p1603135583438600
i've been thinking about doing this for a long time: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8890

Reply:
Fixed in ac33e21